File: core/src/main/java/forpleuvoir/dhwu1a/core/event/message/MessageEvent.java
package forpleuvoir.dhwu1a.core.event.message;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import forpleuvoir.dhwu1a.core.event.base.Dhwu1aEvent;
import forpleuvoir.dhwu1a.core.message.*;
import forpleuvoir.dhwu1a.core.message.base.Message;
import forpleuvoir.dhwu1a.core.user.base.User;
import forpleuvoir.dhwu1a.core.util.ReflectionUtil;

import javax.annotation.Nullable;
import java.util.function.Consumer;

/**
 * 消息事件
 *
 * @author forpleuvoir
 * <p>#project_name dhwu1a
 * <p>#package forpleuvoir.dhwu1a.core.event.message
 * <p>#class_name MessageEvent
 * <p>#create_time 2021/7/4 2:28
 */
public abstract class MessageEvent<M extends Message> extends Dhwu1aEvent {
    /**
     * 所有消息事件都有相关消息
     */
    public final M message;

    protected MessageEvent(M message) {
        this.message = message;
    }

    public static MessageEvent<?> parse(Message message) {
        return switch (message.type) {
            case GroupMessage -> new GroupMessageEvent((GroupMessage) message);
            case FriendMessage -> new FriendMessageEvent((FriendMessage) message);
            case TempMessage -> new TempMessageEvent((TempMessage) message);
            case StrangerMessage -> new StrangerMessageEvent((StrangerMessage) message);
            case OtherClientMessage -> new OtherClientMessageEvent((OtherClientMessage) message);
        };
    }

    @Override
    public void printEventLog() {
        message.printMessageLog();
    }

    @Nullable
    public abstract User getUser();

    @SuppressWarnings({"rawtypes", "unchecked"})
    @Override
    public void broadcastHandle(
            ImmutableMap<Class<? extends Dhwu1aEvent>, ImmutableList<Consumer<? extends Dhwu1aEvent>>> eventListeners
    ) {
        super.broadcastHandle(eventListeners);
        if (ReflectionUtil.isExtended(this.getClass(), MessageEvent.class)) {
            if (eventListeners.containsKey(MessageEvent.class)) {
                for (Consumer listener : eventListeners.get(MessageEvent.class)) {
                    listener.accept(this);
                }
            }
        }
    }

    @Override
    public String toPlainText() {
        return String.format("%s[%s]", message.type, message.toPlainText());
    }

    @Override
    public String toString() {
        return toPlainText();
    }
}

File: core/src/main/kotlin/forpleuvoir/dhwu1a/core/event/message/MessageEvent.kt
package forpleuvoir.dhwu1a.core.event.message

import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import forpleuvoir.dhwu1a.core.event.base.Dhwu1aEvent
import forpleuvoir.dhwu1a.core.message.*
import forpleuvoir.dhwu1a.core.message.base.Message
import forpleuvoir.dhwu1a.core.message.base.MessageType
import forpleuvoir.dhwu1a.core.user.base.User
import forpleuvoir.dhwu1a.core.util.ReflectionUtil.isExtended
import java.util.function.Consumer

/**
 * 消息事件
 * @author forpleuvoir

 * #project_name dhwu1a

 * #package forpleuvoir.dhwu1a.core.event.message

 * #class_name MessageEvent

 * #create_time 2021/8/10 22:14
 */
abstract class MessageEvent<M : Message>(val message: M) : Dhwu1aEvent() {
    companion object {
        @JvmStatic
        @Suppress("UNCHECKED_CAST")
        fun parse(message: Message): MessageEvent<Message> {
            return when (message.type) {
                MessageType.GroupMessage -> GroupMessageEvent(message as GroupMessage) as MessageEvent<Message>
                MessageType.FriendMessage -> FriendMessageEvent(message as FriendMessage) as MessageEvent<Message>
                MessageType.TempMessage -> TempMessageEvent(message as TempMessage) as MessageEvent<Message>
                MessageType.StrangerMessage -> StrangerMessageEvent(message as StrangerMessage) as MessageEvent<Message>
                MessageType.OtherClientMessage -> OtherClientMessageEvent(message as OtherClientMessage) as MessageEvent<Message>
            }
        }
    }

    abstract val user: User?

    override fun printEventLog() {
        message.printMessageLog()
    }

    override fun broadcastHandle(eventListeners: ImmutableMap<Class<out Dhwu1aEvent>, ImmutableList<Consumer<in Dhwu1aEvent>>>) {
        super.broadcastHandle(eventListeners)
        if (isExtended(this.javaClass, MessageEvent::class.java)) {
            if (eventListeners.containsKey(MessageEvent::class.java)) {
                for (listener in eventListeners[MessageEvent::class.java]!!) {
                    listener.accept(this)
                }
            }
        }
    }

    override fun toPlainText(): String {
        return String.format("%s[%s]", message.type, message.toPlainText())
    }

    override fun toString(): String {
        return toPlainText()
    }

}